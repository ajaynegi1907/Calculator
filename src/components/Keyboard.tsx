import React from "react";

export interface KeyboardProps{
    onTokenButtonClick: (token: string) => void;
    clearAll: () => void;
    removePreviousToken: () => void;
    calculateResultClick: () => void;
}

const Keyboard = (props:KeyboardProps) => {
    const onButtonClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void =>{
        const target = event.target as HTMLElement;

        const value = target.innerHTML;
        switch(value){
            case "AC":{
                props.clearAll();
                return;
            }
            case "CE":{
                props.removePreviousToken();
                return;
            }
               
            case "=":{
                props.calculateResultClick();
                return;
            }
            default:{
                props.onTokenButtonClick(value);
            }
        }

    }
    return (
        <div className="keyboard" >
            <button className = "btn btn-action"  onClick={onButtonClick}>AC</button>
            <button className = "btn btn-action"  onClick={onButtonClick}>CE</button>
            <button className = "btn btn-operator"onClick={onButtonClick} >%</button>
            <button className = "btn btn-operator"onClick={onButtonClick} >/</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>7</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>8</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>9</button>
            <button className = "btn btn-operator"onClick={onButtonClick} >*</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>4</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>5</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>6</button>
            <button className = "btn btn-operator"onClick={onButtonClick} >-</button>
            <button className = "btn btn-numeric" onClick={onButtonClick} >1</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>2</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>3</button>
            <button className = "btn btn-operator" onClick={onButtonClick}>+</button>
            <button className = "btn btn-numeric zero" onClick={onButtonClick}>0</button>
            <button className = "btn btn-numeric" onClick={onButtonClick}>.</button>
            <button className = "btn btn-calculate" onClick={onButtonClick}>=</button>
        </div>
    );
}
export default Keyboard;