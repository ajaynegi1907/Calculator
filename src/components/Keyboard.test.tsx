import { shallow, ShallowWrapper } from "enzyme";
import Keyboard, { KeyboardProps } from "./Keyboard";

const props: KeyboardProps = {
    onTokenButtonClick: jest.fn(),
    clearAll: jest.fn(),
    removePreviousToken: jest.fn(),
    calculateResultClick: jest.fn(),
  };
let wrapper: ShallowWrapper;
beforeEach( () => {
     wrapper = shallow(<Keyboard {...props} />);
})

afterEach(() => {
    jest.clearAllMocks();
})

it("should match snapshot", () => {
  expect(wrapper).toMatchSnapshot();
});

it("should call clear all when AC button is clicked", () => {
  const acButton = wrapper.find(".keyboard").childAt(0);
  const mockEvent = {
    target:{
      innerHTML: "AC"
    }
  };
  acButton.simulate("click", mockEvent);
  expect(props.clearAll).toHaveBeenCalled();
});

it("should call calculateResultClick when = button is clicked", () => {
  const calculateButton = wrapper.find(".keyboard").childAt(18);
  const mockEvent = {
    target:{
      innerHTML: "="
    }
  };
  calculateButton.simulate("click", mockEvent);
  expect(props.calculateResultClick).toHaveBeenCalled();
});

it("should call onTokenButtonClick when any token button is clicked", () => {
  const tokenButton = wrapper.find(".keyboard").childAt(4);
  const mockEvent = {
    target:{
      innerHTML: 7
    }
  };
  tokenButton.simulate("click", mockEvent);
  expect(props.onTokenButtonClick).toHaveBeenCalledWith(7);
});

it("should call removePreviousToken when CE button is clicked", () => {
  const keyboard = wrapper.find(".keyboard").childAt(1);
  const mockEvent = {
    target:{
      innerHTML: "CE"
    }
  };
  keyboard.simulate("click", mockEvent);
  expect(props.removePreviousToken).toHaveBeenCalled();
});
