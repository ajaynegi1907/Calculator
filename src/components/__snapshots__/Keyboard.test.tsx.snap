// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="keyboard"
>
  <button
    className="btn btn-action"
    onClick={[Function]}
  >
    AC
  </button>
  <button
    className="btn btn-action"
    onClick={[Function]}
  >
    CE
  </button>
  <button
    className="btn btn-operator"
    onClick={[Function]}
  >
    %
  </button>
  <button
    className="btn btn-operator"
    onClick={[Function]}
  >
    /
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    7
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    8
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    9
  </button>
  <button
    className="btn btn-operator"
    onClick={[Function]}
  >
    *
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    4
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    5
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    6
  </button>
  <button
    className="btn btn-operator"
    onClick={[Function]}
  >
    -
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    1
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    3
  </button>
  <button
    className="btn btn-operator"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="btn btn-numeric zero"
    onClick={[Function]}
  >
    0
  </button>
  <button
    className="btn btn-numeric"
    onClick={[Function]}
  >
    .
  </button>
  <button
    className="btn btn-calculate"
    onClick={[Function]}
  >
    =
  </button>
</div>
`;
